src/IECoreDelight/Renderer.cpp

```diff
@@ -1357,6 +1403,28 @@ class DelightRenderer final : public IECoreScenePreview::Renderer
 
 			updateCamera();
 
+			if( m_evaluatefile != "" )
+			{
+				string evaluatetype = "";
+				if( boost::ends_with( m_evaluatefile, "lua" ) )
+				{
+					evaluatetype = "lua";
+				}
+				else
+				{
+					evaluatetype = "apistream";
+				}
+
+				ParameterList evaluateparams;
+				evaluateparams.add( "type", evaluatetype);
+				evaluateparams.add( "filename", m_evaluatefile);
+
+				NSIEvaluate(
+					m_context,
+					evaluateparams.size(), evaluateparams.data()
+				);
+			}
```

Comment:
This feels awkward from the perspective of interactive rendering, because each time the option is edited, we'll evaluate a new file. So the evaluations will all stack up on top of one another, and we're not seeing what we'd see in a batch render (where only the last evaluation would happen). If what I said makes sense, I'd like to omit this option until we can implement it in a more robust manner.

One thing I've been considering is the possibility for `Renderer::option()` to be allowed to return `false` to say "sorry, I can't edit that now, you'll need to restart the render from scratch". That would trigger the RenderController to restart the render from scratch, so that only the latest file would be evaluated in the new renderer. That would also be useful for Cycles, where there are various things that can't be changed after a render has started.

Reply:
I completely understand what you mean and you are absolutely right that changing the contents of the evaluated file in an interactive session would lead to stacking and incorrect results. Having said that I think that the benefits of having evaluate file available far outweigh the negatives that can happen during interactive rendering. 

For ex. evaluate file can be used to add in a very simple way some 3Delight options that Gaffer doesn't natively support (like lens shaders), it can be used as a world level procedural, etc. Basically it opens to the user flexibility that is only accessible in another way by outputting a NSI ASCII file and then manipulating the NSI file before sending it to the renderer (which can be very slow with anything except the simplest of scenes). 

What I can do to make it obvious to the user that it's dangerous to use evaluate file during interactive rendering is create a Msg:Warning when the option is turned on and the interactive render session is running and also update the tool tip to mention the dangers of using it during an interactive session. Hope this sounds good.

Reply:
(Quoting your comment from https://github.com/GafferHQ/gaffer/pull/5617, because I think it's relevant here).

> I wanted to give a bit of a background to my general approach and reasoning to the Gaffer / 3Delight integration work that I've been doing recently (both this PR and the other ones I submitted the past week or so). Since 3Delight's integration in Gaffer is in many ways in a beta/heavily WIP state, my main goal was to add features that although might not have the most elegant implementation or might not provide optimal user experience, they hopefully enable power users to get as much as possible out of 3Delight within Gaffer. This unfortunately means that until implemented in a better way some things will be somewhat unintuitive or will require a somewhat hacky approach to accomplish the desired result.

While 3Delight support is undoubtedly in an incomplete state, I'm reluctant to use that as a reason for accepting what might be considered temporary solutions, or what can't be implemented robustly for the long run. Once we've released a feature, our general assumption is that people are relying on that in production, and that we'll need to support it potentially for evermore. We _do_ make breaking changes of course, but they are always limited to major versions, and we often need to put considerable time and effort into making the transition. So while I appreciate that exposing `NSIEvaluate` in raw form is useful for power users who are capable of managing the caveats, I don't think that fits with our wider goals or principles.

Reading the NSI docs, it seems that they do have a `procedural` node that uses `NSIEvaluate` internally, and that they correctly handle interactive edits to such a node by removing the nodes it created and re-evaluating. I think we could hook this up via Gaffer's ExternalProcedural node to provide a more compelling and robust integration than a single global option. Do you think that could work?

Reply:
Completely makes sense. I was looking at things more in regards to the short term gain, but you are absolutely right that that might create problems long term. I removed the evaluate file code and I completely agree that down the line implementing NSI procedural node support through Gaffer's ExternalProcedural would be a better way to get the evaluate file benefits and much more.